airflow/models/dag_version.py

```diff
@@ -0,0 +1,163 @@
+        "SerializedDagModel",
+        back_populates="dag_version",
+        uselist=False,
+        cascade="all, delete, delete-orphan",
+        cascade_backrefs=False,
+    )
+    dag_runs = relationship("DagRun", back_populates="dag_version", cascade="all, delete, delete-orphan")
+    task_instances = relationship("TaskInstance", back_populates="dag_version")
+    created_at = Column(UtcDateTime, default=timezone.utcnow)
+
+    __table_args__ = (
+        UniqueConstraint("dag_id", "version_number", name="dag_id_version_number_unique_constraint"),
+    )
+
+    def __repr__(self):
+        return f"<DagVersion {self.dag_id} - {self.version_name}>"
+
+    @classmethod
+    @provide_session
+    def write_dag(
+        cls,
+        *,
+        dag_id: str,
+        version_name: str | None = None,
+        version_number: int = 1,
+        session: Session = NEW_SESSION,
+    ):
+        """Write a new DagVersion into database."""
+        existing_dag_version = session.scalar(
+            with_row_locks(cls._latest_version_select(dag_id), of=DagVersion, session=session, nowait=True)
+        )
+        if existing_dag_version:
+            version_number = existing_dag_version.version_number + 1
+        if existing_dag_version and not version_name:
+            version_name = existing_dag_version.version_name
+
+        dag_version = DagVersion(
+            dag_id=dag_id,
+            version_number=version_number,
+            version_name=version_name,
+        )
+        log.debug("Writing DagVersion %s to the DB", dag_version)
+        session.add(dag_version)
+        # Flush is necessary here due to the unique constraint and other linked tables
+        session.flush()
+        log.debug("DagVersion %s written to the DB", dag_version)
+        return dag_version
+
+    @classmethod
+    def _latest_version_select(cls, dag_id: str):
+        return select(cls).where(cls.dag_id == dag_id).order_by(cls.created_at.desc()).limit(1)
+
+    @classmethod
+    @provide_session
+    def get_latest_version(cls, dag_id: str, session: Session = NEW_SESSION):
+        return session.scalar(cls._latest_version_select(dag_id))
+
+    @classmethod
+    @provide_session
+    def get_version(
+        cls,
+        dag_id: str,
+        version_name: str | None = None,
+        version_number: int | None = None,
+        session: Session = NEW_SESSION,
+    ):
+        version_select_obj = select(cls).where(cls.dag_id == dag_id)
+        if version_name:
+            version_select_obj = version_select_obj.where(cls.version_name == version_name)
+        if version_number:
+            version_select_obj = version_select_obj.where(cls.version_number == version_number)
+        version_select_obj = version_select_obj.order_by(cls.version_number.desc()).limit(1)
+        return session.scalar(version_select_obj)
+
+    @property
+    def version(self):
+        if not self.version_name and not self.version_number:
+            return None
```

Comment:
```suggestion
            return str(self.version_number)
```

If the optional `version_name` isn't set, shouldn't we still return something?

Reply:
Yeah. This condition is wrong because the fields are not nullable. I will fix it